File: topics/mpi/practicals/04.MPI_Type/cxx/1.struct.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the CSCS Summer School.        *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * CSCS take no responsibility for the use of the enclosed      *
 * teaching material.                                           *
 *                                                              *
 * Purpose: Struct derived datatype                             *
 *                                                              *
 * Contents: C-Source                                           *
 *                                                              *
 ****************************************************************/

#include <stdio.h>
#include <mpi.h>

#define NELEM 25

int main(int argc, char *argv[])
{

    typedef struct {
        float x, y, z;
        float velocity;
        int  n, type;
    } Particle;

    int numtasks, rank, source=0, dest, tag=1, i;
    Particle p[NELEM];
    MPI_Datatype particletype, oldtypes[2];
    MPI_Status stat;
    int blockcounts[2];

    /* MPI_Aint type is used to be consistent with syntax of MPI_Type_extent */
    MPI_Aint offsets[2], extent;

    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numtasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    /* Setup description of the 4 MPI_FLOAT fields x, y, z, velocity */

    /* Setup description of the 2 MPI_INT fields n, type */

    /* Now define structured type and commit it */

    /* setup and send particules */
    if (rank == 0) {
        for (i=0; i<NELEM; i++) {
            p[i].x = i * 1.0;
            p[i].y = i * -1.0;
            p[i].z = i * 1.0;
            p[i].velocity = 0.25;
            p[i].n = i;
            p[i].type = i % 2;
        }
    }

    MPI_Bcast(p, NELEM, particletype, 0, MPI_COMM_WORLD);

    printf("rank= %d   %3.2f %3.2f %3.2f %3.2f %d %d\n", rank, p[3].x, p[3].y, p[3].z, p[3].velocity, p[3].n, p[3].type);

    /* Free type */
    MPI_Finalize();
    return 0;
}

File: topics/mpi/practicals/04.MPI_Type/fortran/1.struct.f90
PROGRAM struct

!==============================================================!
!                                                              !
! This file has been written as a sample solution to an        !
! exercise in a course given at the CSCS Summer School.        !
! It is made freely available with the understanding that      !
! every copy of this file must include this header and that    !
! CSCS take no responsibility for the use of the enclosed      !
! teaching material.                                           !
!                                                              !
! Purpose: Struct derived datatype                             !
!                                                              !
! Contents: F-Source                                           !
!==============================================================!

   USE MPI
   IMPLICIT NONE

   INTEGER NELEM
   PARAMETER(NELEM=25)

   TYPE Particle
   SEQUENCE
   REAL*4 x, y, z, velocity
   INTEGER n, type
   END TYPE Particle

   INTEGER numtasks, rank, source, dest, tag, i, ierror
   INTEGER stat(MPI_STATUS_SIZE)
   TYPE (Particle) p(NELEM)
   INTEGER particletype, oldtypes(0:1), blockcounts(0:1), offsets(0:1), extent

   CALL MPI_Init(ierror)
   CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)
   CALL MPI_Comm_size(MPI_COMM_WORLD, numtasks, ierror)

   ! Setup description of the 4 MPI_FLOAT fields x, y, z, velocity

   ! Setup description of the 2 MPI_INT fields n, type
   ! Need to first identify the offset by getting size of MPI_FLOAT

   ! Now define structured type and commit it

   ! Initialize the particle array and then send it to each task
   IF (rank.EQ.0) THEN
      DO i=0, NELEM-1
         p(i) = Particle(1.0*i, -1.0*i, 1.0*i, 0.25, i, mod(i,2))
      END DO
   ENDIF

   CALL MPI_Bcast(p, NELEM, particletype, 0, MPI_COMM_WORLD, ierror)

   WRITE (*,*) "Rank= ", rank, p(3)

   ! Free the type
   CALL MPI_Finalize(ierror)

END PROGRAM
